flow-data/src/main/java/com/vaadin/flow/data/provider/DataCommunicator.java

```diff
@@ -1026,6 +1048,7 @@ private String getInvalidContractMessage(String method) {
     }
 
     private void handleAttach() {
+        ui = UI.getCurrent();
```

Comment:
Using thread local should be avoided - instead use 
```suggestion
        ui = ((StateTree) stateNode.getOwner()).getUI();
```

Reply:
Another option would be to create a method that gets the UI but maybe this is even safer in the long run as it always exists here and doesn't exist after `handleDetach` is called

Reply:
Done